File: adapters/hermes/handler.py
"""birdwatch hook for Hermes — forwards agent:end to the birdwatch dispatcher.

Install: copy this directory to ~/.hermes/hooks/birdwatch/ and restart the
gateway. The dispatcher path defaults to a clone at ~/github/birdwatch;
override with the BIRDWATCH_DISPATCH environment variable.
"""

import asyncio
import json
import os
from pathlib import Path

DISPATCH = os.environ.get(
    "BIRDWATCH_DISPATCH",
    str(Path.home() / "github/birdwatch/scripts/dispatch.sh"),
)


async def handle(event_type, context):
    if event_type != "agent:end":
        return None
    text = (context.get("response") or "").strip()
    if not text:
        return None
    platform = context.get("platform") or "hermes"
    chat = context.get("chat_id") or context.get("user_id") or "chat"
    payload = json.dumps(
        {
            "session_id": str(context.get("session_id") or f"{platform}-{chat}"),
            "hook_event_name": "Stop",
            "cwd": f"hermes/{platform}-{chat}",
            "text": text[:500],
        }
    )
    try:
        proc = await asyncio.create_subprocess_exec(
            "bash",
            DISPATCH,
            "Stop",
            stdin=asyncio.subprocess.PIPE,
            stdout=asyncio.subprocess.DEVNULL,
            stderr=asyncio.subprocess.DEVNULL,
        )
        await asyncio.wait_for(proc.communicate(payload.encode()), timeout=10)
    except Exception:
        pass  # never block the gateway
    return None

File: adapters/openclaw/handler.ts
// birdwatch hook for OpenClaw — forwards outbound agent messages to the
// birdwatch dispatcher as a Stop event so each agent sings as a bird.
// Kept logic-free: extract text/session/project defensively, spawn, detach.
import { spawn } from "node:child_process";
import { homedir } from "node:os";
import { join } from "node:path";

const DISPATCH =
  process.env.BIRDWATCH_DISPATCH ||
  join(homedir(), "github/birdwatch/scripts/dispatch.sh");

const handler = async (event) => {
  if (!event || event.type !== "message" || event.action !== "sent") return;
  const ctx = event.context || event.payload || {};
  const raw = ctx.content ?? ctx.text ?? ctx.body ?? "";
  const text = String(typeof raw === "string" ? raw : JSON.stringify(raw)).trim();
  if (!text) return;
  const session = String(ctx.sessionKey ?? ctx.sessionId ?? "openclaw");
  const project = String(
    ctx.workspaceDir ?? ctx.cwd ?? (ctx.agentId ? `openclaw/${ctx.agentId}` : "openclaw"),
  );
  const payload = JSON.stringify({
    session_id: session,
    hook_event_name: "Stop",
    cwd: project,
    text: text.slice(0, 500),
  });
  try {
    const child = spawn("bash", [DISPATCH, "Stop"], {
      stdio: ["pipe", "ignore", "ignore"],
      detached: true,
    });
    child.stdin.end(payload);
    child.unref();
  } catch {
    // never block the gateway
  }
};

export default handler;
